core/market/src/negotiation/common.rs

```diff
@@ -125,12 +132,61 @@ impl CommonBroker {
         Ok((new_proposal, is_first))
     }
 
+    pub async fn reject_proposal(
+        &self,
+        subs_id: Option<&SubscriptionId>,
+        proposal_id: &ProposalId,
+        caller_id: &NodeId,
+        caller_role: Owner,
+        reason: &Option<Reason>,
+    ) -> Result<Proposal, RejectProposalError> {
+        // Check if subscription is still active.
+        // Note that subscription can be unsubscribed, before we get to saving
+        // Proposal to database. This seems like race conditions, but there's no
+        // danger of data inconsistency. If we won't reject countering Proposal here,
+        // it will be sent to Provider and his counter Proposal will be rejected later.
+
+        let proposal = self.get_proposal(subs_id.clone(), proposal_id).await?;
+
+        self.validate_proposal(&proposal, caller_id, caller_role)
+            .await?;
+
+        if proposal.body.issuer == Issuer::Us && proposal_id.owner() == caller_role {
+            let e = ProposalValidationError::OwnProposal(proposal.body.id.clone());
+            log::warn!("{}", e);
+            Err(e)?;
+        }
+
+        self.db
+            .as_dao::<ProposalDao>()
+            .change_proposal_state(proposal_id, ProposalState::Rejected)
+            .await?;
+
+        match caller_role {
+            Owner::Requestor => counter!("market.proposals.requestor.rejected", 1),
+            Owner::Provider => counter!("market.proposals.provider.rejected", 1),
+        };
```

Comment:
**Important to consider:** I'm wondering, that maybe we should distinguish between proposals that we reject as Provider/Requestor and ours proposals that are being rejected.
Now you use this function in both contexts

Reply:
fixed